pkg/gameservers/health_test.go

```diff
@@ -79,15 +79,114 @@ func TestHealthUnschedulableWithNoFreePorts(t *testing.T) {
 	assert.False(t, hc.unschedulableWithNoFreePorts(pod))
 }
 
+func TestHealthControllerSkipUnhealthy(t *testing.T) {
+	t.Parallel()
+
+	fixtures := map[string]struct {
+		setup    func(*agonesv1.GameServer, *corev1.Pod)
+		expected bool
+	}{
+		"scheduled and terminated container": {
+			setup: func(gs *agonesv1.GameServer, pod *corev1.Pod) {
+				gs.Status.State = agonesv1.GameServerStateScheduled
+				pod.Status.ContainerStatuses = []corev1.ContainerStatus{{
+					Name:  gs.Spec.Container,
+					State: corev1.ContainerState{Terminated: &corev1.ContainerStateTerminated{}},
+				}}
+			}, expected: true},
+		"after ready and terminated container": {
+			setup: func(gs *agonesv1.GameServer, pod *corev1.Pod) {
+				gs.Status.State = agonesv1.GameServerStateReady
+				pod.Status.ContainerStatuses = []corev1.ContainerStatus{{
+					Name:  gs.Spec.Container,
+					State: corev1.ContainerState{Terminated: &corev1.ContainerStateTerminated{}},
+				}}
+			},
+			expected: false,
+		},
+		"before ready, with no terminated container": {
+			setup: func(gs *agonesv1.GameServer, pod *corev1.Pod) {
+				gs.Status.State = agonesv1.GameServerStateScheduled
```

Comment:
the before ready states are all the same, given the other test coverage it may not matter but you could vary which state is set to one of the other 4 states that are before ready.